prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+	return FI_SUCCESS;
+}
+
+
+static int tcpx_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+	struct tcpx_ep *ep;
+
+	ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid.fid);
+	switch (command) {
+	case FI_ENABLE:
+		if (!ep->util_ep.rx_cq || !ep->util_ep.tx_cq)
+			return -FI_ENOCQ;
+		break;
+	default:
+		return -FI_ENOSYS;
+	}
+	return 0;
+}
+
+static struct fi_ops tcpx_ep_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = tcpx_ep_close,
+	.bind = tcpx_ep_bind,
+	.control = tcpx_ep_ctrl,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_ep tcpx_ep_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = fi_no_cancel,
+	.getopt = fi_no_getopt, //tcpx_getopt,
+	.setopt = fi_no_setopt, //tcpx_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void tcpx_manual_progress(struct util_ep *util_ep)
+{
+	return;
+}
+
+int tcpx_endpoint(struct fid_domain *domain, struct fi_info *info,
+		  struct fid_ep **ep_fid, void *context)
+{
+	struct tcpx_ep *ep;
+	struct tcpx_conn_handle *handle;
+	int ret;
+
+	ep = calloc(1, sizeof(*ep));
+	if (!ep)
+		return -FI_ENOMEM;
+
+	ret = ofi_endpoint_init(domain, &tcpx_util_prov, info, &ep->util_ep,
+				context, tcpx_manual_progress);
+	if (ret)
+		goto err;
+
+	ep->info = *info;
+	ofi_atomic_initialize32(&ep->ref, 0);
+
+	if (info != NULL && info->handle != NULL){
+		handle = container_of(info->handle,struct tcpx_conn_handle,
+				      handle);
+		ep->handle = handle;
+		fi_fd_nonblock(handle->conn_fd);
+	} else {
+		ep->handle = calloc(1, sizeof(*handle));
+		if (!ep->handle) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+	}
+
+	*ep_fid = &ep->util_ep.ep_fid;
+	(*ep_fid)->fid.ops = &tcpx_ep_fi_ops;
+	(*ep_fid)->ops = &tcpx_ep_ops;
+	(*ep_fid)->cm = &tcpx_cm_ops;
+	(*ep_fid)->msg = &tcpx_msg_ops;
+
+	return 0;
+err:
+	free(ep);
+	return ret;
+}
+
+static int tcpx_pep_fi_close(struct fid *fid)
+{
+	struct tcpx_pep *pep;
+	int ret;
+	char c;
+
+	pep = container_of(fid, struct tcpx_pep, pep.fid);
+	pep->cm.do_listen = 0;
+
+	ret = ofi_write_socket(pep->cm.signal_fds[0], &c, 1);
+	if (ret != 1)
+		FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,"Failed to signal\n");
+
+	if (pep->cm.listener_thread &&
+	    pthread_join(pep->cm.listener_thread, NULL)) {
+		FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,"pthread join failed\n");
+	}
+
+	ofi_close_socket(pep->cm.signal_fds[0]);
+	ofi_close_socket(pep->cm.signal_fds[1]);
+
+	free(pep);
+	return 0;
+
+}
+
+static int tcpx_pep_fi_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
+{
+	struct tcpx_pep *pep;
+	struct util_eq *eq;
+
+	pep = container_of(fid, struct tcpx_pep, pep.fid);
+
+	if (bfid->fclass != FI_CLASS_EQ)
+		return -FI_EINVAL;
+
+	eq = container_of(bfid, struct util_eq, eq_fid.fid);
+	if (pep->fabric != eq->fabric) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"Cannot bind Passive EP and EQ on different fabric\n");
+		return -FI_EINVAL;
+	}
+	pep->eq = eq;
+	return 0;
+}
```

Comment:
See about moving all or portions of the pep close and bind calls into common code, with a util_pep structure.